### ISSUE TITLE
🚀 Contributing to Keras: Open Projects & Ideas 🚀

### ISSUE THREAD

**Issue by fchollet** [keras-team/keras#18442]
Keras 3 is a high-velocity open-source project. We welcome contributions! Below are some ideas on how to get started. For specific contributing steps, please see our [CONTRIBUTING.md](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md).

## Areas we're looking for help in

- Search for issues with [`label:"Good first issue"`](https://github.com/keras-team/keras/issues?q=state%3Aopen%20label%3A%22Good%20first%20issue%22) or [`label:"stat:contributions welcome"`](https://github.com/keras-team/keras/issues?q=state%3Aopen%20label%3A%22stat%3Acontributions%20welcome%22)
- Performance optimization
  - Improve `torch` backend performance
  - Add and optimize [image preprocessing layers](https://github.com/keras-team/keras/issues/23245#issuecomment-5058025264)
- Contribute to the MLX backend. We have a `mlx` branch on the repo, which does not yet have full operation convergence in `keras/src/backend/mlx/`. Add missing ops and help get the new MLX backend closer to the finish line!

## Bring new pretrained model implementations to KerasHub

The Keras ecosystem package [KerasHub](https://github.com/keras-team/keras-hub) is looking for contributors!
If you'd like to port a pretrained model implementation to the Keras ecosystem, that's a great project to join.

## Add new examples to keras.io

Have a great ML project or technique you'd like to use to teach others about ML? You can add a new tutorial on [keras.io/examples](https://keras.io/examples/)!

We are especially interested in new examples covering modern architectures and workflows. Check out our **[Call for Code Example Contributions](https://github.com/keras-team/keras-io/blob/master/call_for_contributions.md)** for an active list of tutorial ideas we're looking for (such as DETR object detection, text-to-image diffusion, MIDI music generation, and structured data with Keras Preprocessing Layers).

To submit a tutorial, follow the contribution instructions at [keras-team/keras-io](https://github.com/keras-team/keras-io).

## Increase test coverage

Keras doesn't have 100% code coverage in unit tests just yet. You can try to run a code coverage tool like `codecov` and see if there are any unit tests you could add to improve Keras.

**Comment by GreatRSingh** [comment]
@fchollet I am interested in solving the todo,
keras/metrics/accuracy_metrics_test.py:        # TODO: Check save and restore config

can you give me more detail of how it should be done.

**Comment by naoitoi** [comment]
Hi, @fchollet , 

For my first commit to Keras, I picked a simple one, addressing two small TODO items in reduction_metrics_test.py.

Here's the PR: https://github.com/keras-team/keras/pull/18812

I believe what the comment "# TODO: Check save and restore config" meant is that an object (Sum or Mean) must be re-constructed from config, and be tested.

Please let me know if I misunderstood the requirements, or otherwise made any mistakes.

Thank you!  I'm very excited about working on this very important library.

**Comment by AakashKumarNain** [comment]
Is anyone working on JAX CTC? If not, I can take it up

**Comment by MaanasArora** [comment]
Hi, I'd like to get started on the TODO referenced in some of the `model_test.py` tests:

https://github.com/keras-team/keras/blob/fe2f54aa5bc42fb23a96449cf90434ab9bb6a2cd/keras/models/model_test.py#L222

I assume that would require modifying the loss tracker. Let me know if that would be good or if anyone else is working on it, and I can open a PR!

**Comment by said-ml** [comment]
can we add a feature that suggest for the user to use a TPU or GPU if number of parameters of model surpass some threshold?
example:
import tensorflow as tf
from tensorflow import keras
from typing import Union

def suggest_accelerator(model, parameter_threshold=1000000):
  """Suggests using TPU or GPU if model size exceeds a threshold.

  Args:
    model: The Keras model object.
    parameter_threshold: The threshold for number of parameters (int).
  """
  assert isinstance(model, keras.Model)
  num_params = model.count_params()

  if num_params > parameter_threshold:
    if tf.config.list_physical_devices('TPU'):
      print("Your model has", num_params, "parameters, exceeding the threshold of", parameter_threshold)
      print("Consider using a TPU for faster training. See https://cloud.google.com/tpu/docs/ for instructions.")
    elif tf.config.list_physical_devices('GPU'):
      print("Your model has", num_params, "parameters, exceeding the threshold of", parameter_threshold)
      print("Consider using a GPU for faster training. Ensure you have a compatible GPU and enable it in your code.")
    else:
      print("Your model has", num_params, "parameters, exceeding the threshold of", parameter_threshold)
      print("Training on CPU might be slow. Consider using cloud TPUs or a machine with a GPU for faster training.")

# Example usage:
model = tf.keras.Sequential(...)  # Define your model
suggest_accelerator(model)

**Comment by said-ml** [comment]
can we add a feature that suggest for the user to use a TPU or GPU if number of parameters of model surpass some threshold?
import tensorflow as tf
from tensorflow import keras

def suggest_accelerator(model, parameter_threshold=1000000):
  """Suggests using TPU or GPU if model size exceeds a threshold.

  Args:
    model: The Keras model object.
    parameter_threshold: The threshold for number of parameters (int).
  """

  num_params = model.count_params()

  if num_params > parameter_threshold:
    if tf.config.list_physical_devices('TPU'):
      print("Your model has", num_params, "parameters, exceeding the threshold of", parameter_threshold)
      print("Consider using a TPU for faster training. you need more line code , check out the following link 
                     https://www.kaggle.com/code/ryanholbrook/getting-started-with-tpus")
    elif tf.config.list_physical_devices('GPU'):
      print("Your model has", num_params, "parameters, exceeding the threshold of", parameter_threshold)
      print("Consider using a GPU for faster training. Ensure you have a compatible GPU and enable it in your code.")
    else:
      print("Your model has", num_params, "parameters, exceeding the threshold of", parameter_threshold)
      print("Training on CPU might be slow. Consider using cloud TPUs or a machine with a GPU for faster training.")

# Example usage:
model = tf.keras.Sequential(...)  # Define your model
suggest_accelerator(model)

**Comment by vulkomilev** [comment]
I will take the ops.correlate if nobody is working on it.

**Comment by markodjordjic** [comment]
@fchollet I could take dice loss if nobody else is working on it. @vulkomilev did you start with ops.correlate?

**Comment by markodjordjic** [comment]
@fchollet can you or someone add me as a contributor? I need to make a push to the remote branch for the dice loss.

**Comment by markodjordjic** [comment]
@fchollet I have the first version of the DiceLoss in the tf backend. I need to be made a contributor to make a push to the remote branch so that the code review can take place.

**Comment by markodjordjic** [comment]
@fchollet any updates for the DiceLoss? I have the TF implementation ready.

**Comment by vulkomilev** [comment]
> @fchollet I could take dice loss if nobody else is working on it. @vulkomilev did you start with ops.correlate?

yes I did

**Comment by markodjordjic** [comment]
@vulkomilev Thanks for the answer. Are you a contributor? How do you do the push to the branch? I have this dice loss done, but cannot push since I am not a contributor.

**Comment by Frnckthewoo** [comment]
Please Can I have a french version !?

**Comment by YagaoDirac** [comment]
@fchollet Hi, I think I invented something interesting. I implemented it in pytorch and I opened an issue there. 
https://github.com/pytorch/pytorch/issues/122680

**Comment by vulkomilev** [comment]
> @vulkomilev How do you do the push to the branch? 

you make a MR and waiting for approval

**Comment by markodjordjic** [comment]
@vulkomilev I cannot push. I don't have the rights.

**Comment by vulkomilev** [comment]
> @vulkomilev I cannot push. I don't have the rights.

sorry I meant pull request

**Comment by pmasousa** [comment]
Hello @fchollet,
Can I make the tool for saved Keras model file inspection, diff, and patching or is there someone already working on it?
Thanks in advance!

**Comment by vulkomilev** [comment]
I will get the "Add a new Flash Attention op" task

**Comment by vulkomilev** [comment]
@fchollet for every back-end I must introduce the implementation mentioned in the ticket right ?

**Comment by chanduvkp** [comment]
@sampathweb @fran6co - I have seen last comments on file '**keras/src/layers/preprocessing/string_lookup_test.py**' . I have grep'ed for 'TODO' and list this file . I am new for Keras contrib and setup Keras in local successfully. Would any one give me insight on this file and changes needed for 'TODO' sections .

**Comment by PatReis** [comment]
Hello @fchollet ,
I just saw that keras is merging NLP and CV into a models repo. 
If there is interest from the keras team to also include Graph Neural Networks, I am eager to contribute or become involved.
I reworked [kgcnn](https://github.com/aimat-lab/gcnn_keras) for keras 3 and it works nicely, although I would have removed lots of the clutter but I tried to stay backward compatible.
Of course, I would do things differently and much cleaner when starting over!
In my opinion, it is important to be compatible with [PyG](https://github.com/pyg-team/pytorch_geometric) and the disjoint graph representation for keras models as well, so that people can use their input pipelines. 
Does keras has a different strategy? What do you think?

**Comment by ghsanti** [comment]
Hi @fchollet,

Is there any intentions to add types (type information) to this library?

**Comment by VaishnaviMudaliar** [comment]
Hi @fchollet ! I'm interested in contributing to the MLX backend. I noticed the mlx branch appears to be significantly behind master, so I wanted to check whether contributions to this branch are still welcome. If so, are there specific missing operations or files (e.g., numpy.py, math.py, linalg.py) that you'd recommend new contributors start with?

**Comment by kanaad-lims** [comment]
Hi @fchollet ! I built [Torch Transform Inspector](https://github.com/kanaad-lims/Torch-Transform-Inspector), a small library that applies torchvision image transforms and visualizes the output of each transform together in a Matplotlib grid, making augmentation pipelines easier to inspect. I was wondering if a similar tool would be useful for Keras preprocessing/augmentation layers, and if so, I’d be interested in implementing it for Keras.
Thank you!

**Comment by divyashreepathihalli** [comment]
@kanaad-lims That sounds like a wonderful idea. For special feature projects based on Keras, we recommend that you create the tool in your personal profile, and we can help highlight it during our monthly community meetings.

**Comment by kanaad-lims** [comment]
Thank you!! @divyashreepathihalli 
I really appreciate the encouragement. I’ll go ahead and develop it as a standalone project in my profile, and I’ll share the initial implementation here once I have it ready.
